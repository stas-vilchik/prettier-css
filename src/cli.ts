import { readFileSync, writeFileSync } from 'fs';
import { extname } from 'path';
import { sync } from 'glob';
import { format } from './index';

const filePatterns = process.argv[2];

sync(filePatterns).forEach((filename: string) => {
  let text;
  let extension;

  try {
    text = readFileSync(filename, 'utf-8');
    extension = extname(filename);
  } catch (error) {
    console.error(`Unable to read file: "${filename}":\n${error}`);
    process.exitCode = 2;
    return;
  }

  let formatted: string;
  try {
    formatted = format(text, extension.substr(1));
  } catch (error) {
    console.error(`Unable to format file: "${filename}":\n${error}`);
    process.exitCode = 2;
    return;
  }

  try {
    writeFileSync(filename, formatted, 'utf-8');
  } catch (error) {
    console.error(`Unable to write file: "${filename}":\n${error}`);
    process.exitCode = 2;
  }
});