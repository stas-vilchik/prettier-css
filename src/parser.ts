import * as postcss from 'postcss';
const lessSyntax = require('postcss-less');

export function parse(text: string, syntax: string) {
  const result = syntax === 'less'
    ? postcss().process(text, { syntax: lessSyntax })
    : postcss().process(text);
  return result.root;
}
