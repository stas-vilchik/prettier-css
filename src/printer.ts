import { Root, Rule, Declaration, Comment, AtRule, Node } from 'postcss';

const INDENT = '  ';

function indent(str: string, indentLevel: number) {
  return `${Array(indentLevel + 1).join(INDENT)}${str}`;
}

function lineBreaks(str: string = '', initial: number = 0) {
  let count = initial + str.split('').reduce((p, c) => p + (c === '\n' ? 1 : 0), 0);
  return count > 1 ? Array(count).join('\n') : '';
}

function printDeclaration(decl: Declaration, indentLevel: number) {
  let output = `${decl.prop.toLowerCase()}: ${decl.value}`;

  if (decl.important) {
    output += ' !important';
  }

  output += ';';

  return lineBreaks(decl.raws.before) + indent(output, indentLevel) + lineBreaks(decl.raws.after);
}

function printComment(comment: Comment, indentLevel: number) {
  return (
    lineBreaks(comment.raws.before) +
    indent(`/* ${comment.text} */`, indentLevel) +
    lineBreaks(comment.raws.after)
  );
}

function printBlockComment(comment: Comment, indentLevel: number) {
  const lines = [
    '/*',
    ' ' + comment.text,
    ' */',
    ''
  ]
    .map(line => indent(line, indentLevel))
    .join('\n');
  return lineBreaks(comment.raws.before, -1) + lines + lineBreaks(comment.raws.after);
}

function printRule(rule: Rule, indentLevel: number) {
  let output = '';

  const emptyRule = rule.nodes == null || rule.nodes.length === 0;
  const selectors = rule.selector.split(',').map(s => indent(s.trim(), indentLevel)).join(',\n');

  output += `${selectors} ${emptyRule ? "{}" : "{"}\n`;

  (rule.nodes || []).forEach(node => {
    switch (node.type) {
      case 'decl':
        output += printDeclaration(node as Declaration, indentLevel + 1);
        break;
      case 'comment':
        output += printComment(node as Comment, indentLevel + 1);
        break;
      default:
      // console.log(node.type, node);
    }
    output += '\n';
  });

  if (!emptyRule) {
    output += indent(`}\n`, indentLevel);
  }

  return output;
}

function printAtRule(atRule: AtRule, indentLevel: number) {
  let output = indent(`@${atRule.name}`, indentLevel);
  if (atRule.params) {
    output += ' ' + atRule.params;
  }

  const { nodes } = atRule;
  if (nodes) {
    output += ' {\n';
  } else {
    output += ';';
  }

  if (nodes) {
    nodes.forEach((node, index) => {
      switch (node.type) {
        case 'rule':
          output += printRule(node as Rule, indentLevel + 1);
          break;
        case 'decl':
          output += printDeclaration(node as Declaration, indentLevel + 1);
          if (index === nodes.length - 1) {
            output += `\n`;
          }
          break;
        default:
        // console.log(node.type, node);
      }
      if (index < nodes.length - 1) {
        output += `\n`;
      }
    });
  }

  if (atRule.nodes) {
    output += indent(`}\n`, indentLevel);
  } else {
    output += '\n';
  }

  return output;
}

export function print(tree: Root) {
  let output = '';

  const { nodes } = tree;
  if (nodes) {
    nodes.forEach((node, nodeIndex) => {
      switch (node.type) {
        case 'rule':
          output += printRule(node as Rule, 0);
          break;
        case 'atrule':
          output += printAtRule(node as AtRule, 0);
          break;
        case 'comment':
          output += printBlockComment(node as Comment, 0);
          break;
        default:
        // console.log(node.type, node);
      }

      if (nodeIndex < nodes.length - 1) {
        output += `\n`;
      }
    });
  }

  return output;
}