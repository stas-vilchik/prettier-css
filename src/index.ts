import { parse } from './parser';
import { print } from './printer';

export function format(text: string, syntax: string = 'css') {
  const tree = parse(text, syntax);
  return print(tree);
}