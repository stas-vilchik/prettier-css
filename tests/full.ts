import { readFileSync } from 'fs';
import { join } from 'path';
import { format } from '../src/index';

it('selectors', () => {
  const text = readFileSync(join(__dirname, 'selectors.css'), 'utf-8');
  expect(format(text)).toMatchSnapshot();
});

it('rules', () => {
  const text = readFileSync(join(__dirname, 'rules.css'), 'utf-8');
  expect(format(text)).toMatchSnapshot();
});

it('atRules', () => {
  const text = readFileSync(join(__dirname, 'atRules.css'), 'utf-8');
  expect(format(text)).toMatchSnapshot();
});

it('comments', () => {
  const text = readFileSync(join(__dirname, 'comments.css'), 'utf-8');
  expect(format(text)).toMatchSnapshot();
});

it('less', () => {
  const text = readFileSync(join(__dirname, 'variables.less'), 'utf-8');
  expect(format(text, 'less')).toMatchSnapshot();
})