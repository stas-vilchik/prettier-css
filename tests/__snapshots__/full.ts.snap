// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atRules 1`] = `
"@foo a {
  color: red;
}

@foo b {
  color: red;
  text-align: center;
}

@foo c {
  .bar,
  .baz {
    color: red;
  }
}
"
`;

exports[`comments 1`] = `
".foo {
  /* the color is red */
  color: red;
}

.foo {
  /* the color is red */

  color: red;
}

.foo {
  color: red;
  /* the color is red */
}

.foo {
  color: red;
  /* the color is red */
}

.foo {
  color: red;

  /* the color is red */
}

.foo {
  color: red;

  /* the color is red */
}

.foo {
  color: red;
  /* text to the right */
  text-align: right;
}

.foo {
  color: red;
  /* text to the right */

  text-align: right;
}

.foo {
  color: red;

  /* text to the right */
  text-align: right;
}
"
`;

exports[`rules 1`] = `
".a {}

.b {
  color: red;
}
"
`;

exports[`selectors 1`] = `
".a,
.b {
  color: red;
}

.a > .b {
  color: red;
}
"
`;
